activation/nipost_test.go

```diff
@@ -41,6 +48,337 @@ func defaultLayerClockMock(ctrl *gomock.Controller) *MocklayerClock {
+			func(context.Context, []byte) (*types.Post, *types.PostInfo, error) {
+				expectedEvents <- func(e events.UserEvent) {
+					event := e.Event.GetPostComplete()
+					if !assert.NotNil(t, event, "wrong event type") {
+						return
+					}
+
+					assert.Equal(t, true, e.Event.Failure)
+					assert.Equal(t, "Node failed PoST execution.", e.Event.Help)
+				}
+
+				return nil, nil, expectedErr
+			})
+
+		nipost, nipostInfo, err := nb.Proof(context.Background(), shared.ZeroChallenge)
+		require.ErrorIs(t, err, expectedErr)
+		require.Nil(t, nipost)
+		require.Nil(t, nipostInfo)
+	})
+
+	t.Run("repeated connection failure", func(t *testing.T) {
+		t.Parallel()
+
+		sig, err := signing.NewEdSigner()
+		require.NoError(t, err)
+
+		ctrl := gomock.NewController(t)
+		poetDb := NewMockpoetDbAPI(ctrl)
+		mclock := defaultLayerClockMock(ctrl)
+		postService := NewMockpostService(ctrl)
+
+		observer, observedLogs := observer.New(zapcore.WarnLevel)
+		logger := zap.New(observer)
+
+		nb, err := NewNIPostBuilder(
+			localsql.InMemory(),
+			poetDb,
+			postService,
+			[]types.PoetServer{},
+			logger,
+			sig,
+			PoetConfig{},
+			mclock,
+		)
+		require.NoError(t, err)
+
+		setupEventListener(t) // no events
+
+		ctx, cancel := context.WithCancel(context.Background())
+		postService.EXPECT().Client(sig.NodeID()).Return(nil, ErrPostClientNotConnected).Times(10)
+		postService.EXPECT().Client(sig.NodeID()).DoAndReturn(
+			func(types.NodeID) (PostClient, error) {
+				cancel()
+				return nil, ErrPostClientNotConnected
+			}).Times(1)
+
+		nipost, nipostInfo, err := nb.Proof(ctx, shared.ZeroChallenge)
+		require.ErrorIs(t, err, context.Canceled)
+		require.Nil(t, nipost)
+		require.Nil(t, nipostInfo)
+
+		require.Equal(t, 1, observedLogs.Len(), "expected 1 log message")
+		require.Equal(t, zapcore.WarnLevel, observedLogs.All()[0].Level)
+		require.Equal(t, "post service not connected - waiting for reconnection", observedLogs.All()[0].Message)
+		require.Equal(t, sig.NodeID().String(), observedLogs.All()[0].ContextMap()["service id"])
+	})
+}
+
+func setupEventListener(t *testing.T) chan<- func(events.UserEvent) {
+	events.InitializeReporter()
+	sub, _, err := events.SubscribeUserEvents(events.WithBuffer(10))
+	require.NoError(t, err)
+
+	expectedEvents := make(chan func(events.UserEvent), 10)
+	var eg errgroup.Group
+	eg.Go(func() error {
+		for event := range sub.Out() {
+			select {
+			case ev, ok := <-expectedEvents:
+				if !ok {
+					assert.Fail(t, "unexpected event: %v", event)
+					return nil
+				}
+				ev(event)
+			case <-time.After(5 * time.Second):
+				assert.Fail(t, "timeout waiting for event")
+				return nil
+			}
+		}
+		_, ok := <-expectedEvents
+		assert.False(t, ok, "expected more events")
+		return nil
+	})
+	t.Cleanup(func() { eg.Wait() })
+	t.Cleanup(func() { sub.Close() })
+	t.Cleanup(func() {
+		close(expectedEvents)
+		time.Sleep(100 * time.Millisecond) // give subscription some time to handle all events
```

Comment:
100ms is arbitrary and might not be enough on the less stable runners

Reply:
Because the event handling runs over multiple go routines via channels I do have to add some timeouts or I might run into races. This specific one I removed, because it is not needed any more. I now have 5 seconds as timeout everywhere, which should be fast enough even on the slowest runners.